components/supervisor/pkg/ports/ports-config.go

```diff
@@ -0,0 +1,226 @@
+		for port, config := range configs {
+			_, exists := visited[port]
+			if exists {
+				continue
+			}
+			visited[port] = struct{}{}
+			callback(port, config)
+		}
+	}
+}
+
+// Get returns the config for the give port
+func (configs *Configs) Get(port uint32) (*gitpod.PortConfig, bool) {
+	if configs == nil {
+		return nil, false
+	}
+	config, exists := configs.instancePortConfigs[port]
+	if exists {
+		return config, true
+	}
+	config, exists = configs.workspaceConfigs[port]
+	if exists {
+		return config, true
+	}
+	return nil, false
+}
+
+// GetFromRange returns the port config from the port range config
+func (configs *Configs) GetFromRange(port uint32) (*gitpod.PortConfig, bool) {
+	if configs == nil {
+		return nil, false
+	}
+	for _, rangeConfig := range configs.instanceRangeConfigs {
+		if rangeConfig.Start <= port && port <= rangeConfig.End {
+			return &gitpod.PortConfig{
+				Port:       float64(port),
+				OnOpen:     rangeConfig.OnOpen,
+				Visibility: rangeConfig.Visibility,
+			}, true
+		}
+	}
+	return nil, false
+}
+
+// ConfigInterace allows to watch port configurations
+type ConfigInterace interface {
+	// Observe provides channels triggered whenever the port configurations are changed.
+	Observe(ctx context.Context) (<-chan *Configs, <-chan error)
+}
+
+// ConfigService allows to watch port configurations
+type ConfigService struct {
+	workspaceID   string
+	configService gitpod.ConfigInterface
+	gitpodAPI     gitpod.APIInterface
+	parser        *configParser
+}
+
+// NewConfigService creates a new instance of ConfigService
+func NewConfigService(workspaceID string, configService gitpod.ConfigInterface, gitpodAPI gitpod.APIInterface) *ConfigService {
+	return &ConfigService{
+		workspaceID:   workspaceID,
+		configService: configService,
+		gitpodAPI:     gitpodAPI,
+		parser:        newConfigParser(),
+	}
+}
+
+// Observe provides channels triggered whenever the port configurations are changed.
+func (service *ConfigService) Observe(ctx context.Context) (<-chan *Configs, <-chan error) {
+	updatesChan := make(chan *Configs)
+	errorsChan := make(chan error, 1)
+
+	go func() {
+		defer close(updatesChan)
+		defer close(errorsChan)
+
+		configs, errs := service.configService.Observe(ctx)
+
+		current := &Configs{}
+		if service.gitpodAPI != nil {
+			info, err := service.gitpodAPI.GetWorkspace(ctx, service.workspaceID)
+			if err != nil {
+				errorsChan <- err
+			} else {
+				current.workspaceConfigs = service.parser.parseWorkspaceConfigs(info.Workspace.Config.Ports)
+				updatesChan <- &Configs{workspaceConfigs: current.workspaceConfigs}
+			}
+		} else {
+			errorsChan <- errors.New("could not connect to Gitpod API to fetch workspace port configs")
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-errs:
+				errorsChan <- err
+			case config := <-configs:
+				if service.update(config, current) {
```

Comment:
I had to look into `update` to understand what's happening here, in particular what the `true` return means.
Maybe a named return value, or a function comment on `update` would be helpful.

Nit: happy path is not in line of sight (https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88)
```suggestion
if !service.update(config, current) {
    continue
}
```